### DISCUSSION TITLE
MIMXRT1xxx: Supporting 8-channel I2S input via SAI1

### DISCUSSION THREAD

**Discussion by ned-pcs** [micropython/micropython#11603] [m.iMX RT / Teensy 4.x]
The current MicroPython support for I2S on the MIMXRT chips only uses a single data line, though the SAI1 peripheral on the MIMXRT1021 supports up to four lines. I'd like to make an 8 microphone array using I2S MEMS microphones and would love to use MicroPython.

Perhaps for I2S port 1 the `sd` argument could take either a single pin (as it does now), or a tuple with one to four pins. For example (I have made up the pin numbers here):

```python
i2s = I2S(1, sck=Pin(21), ws=Pin(20), sd=(Pin(7), Pin(8), Pin(9), Pin(10)), mck=Pin(23),
    mode=I2S.RX, bits=32, rate=48000,format=I2S.STEREO,
    ibuf=40000,
)
```

The `RDR0-RDR3` receive data registers (as well as the `TDR0-TDR3` transmit data registers) are at sequential word addresses, so DMA could be configured to handle this easily.

Is anyone working on this? Would supporting this feature in the way I have proposed above be incompatible with MicroPython design goals?

Thanks!

**Comment by miketeachman** [comment]
I'm curious on the application for the microphone array.  How would the microphone audio data be used?

**Comment by ned-pcs** [comment]
We're doing some experiments with Tensorflow and direction of arrival detection. One application is to distinguish and characterize different noise sources within an environment. We've been using a small MEMS microphone array that plugs into a RPi that has 8 I2S microphones, but that doesn't allow for easy physical re-configuration of the microphones.
